exporter/sapmexporter/correlation.go

```diff
@@ -0,0 +1,146 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sapmexporter
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"sync"
+
+	"github.com/signalfx/signalfx-agent/pkg/apm/correlations"
+	"github.com/signalfx/signalfx-agent/pkg/apm/tracetracker"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/common/splunk"
+)
+
+// Tracker correlation
+type Tracker struct {
+	once         sync.Once
+	log          *zap.Logger
+	cfg          *Config
+	params       component.ExporterCreateParams
+	traceTracker *tracetracker.ActiveServiceTracker
+	correlation  *correlationContext
+}
+
+type correlationContext struct {
+	correlations.CorrelationClient
+	cancel context.CancelFunc
+}
+
+// NewTracker creates a new tracker instance for correlation.
+func NewTracker(cfg *Config, params component.ExporterCreateParams) *Tracker {
+	return &Tracker{
+		log:    params.Logger,
+		cfg:    cfg,
+		params: params,
+	}
+}
+
+func newCorrelationClient(cfg *Config, params component.ExporterCreateParams) (
+	*correlationContext, error,
+) {
+	endpoint := cfg.APIEndpoint
+	if cfg.Correlation.Endpoint != "" {
+		endpoint = cfg.Correlation.Endpoint
+	}
+
+	corrURL, err := url.Parse(endpoint)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse correlation endpoint URL %q: %v", cfg.Correlation.Endpoint, err)
+	}
+
+	httpClient, err := cfg.Correlation.ToClient()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create correlation API client: %v", err)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	client, err := correlations.NewCorrelationClient(newZapShim(params.Logger), ctx, httpClient, correlations.ClientConfig{
+		Config:      cfg.Correlation.Config,
+		AccessToken: cfg.AccessToken,
+		URL:         corrURL,
+	})
+
+	if err != nil {
+		cancel()
+		return nil, fmt.Errorf("failed to create correlation client: %v", err)
+	}
+
+	return &correlationContext{
+		CorrelationClient: client,
+		cancel:            cancel,
+	}, nil
+}
+
+// AddSpans processes the provided spans to correlate the services and environment observed
+// to the resources (host, pods, etc.) emitting the spans.
+func (cor *Tracker) AddSpans(ctx context.Context, traces pdata.Traces) {
+	if traces.ResourceSpans().Len() == 0 {
+		return
+	}
+
+	cor.once.Do(func() {
+		if !cor.cfg.Correlation.Enabled {
+			cor.log.Info("Span/metrics correlation is disabled")
+			return
+		}
+
+		res := traces.ResourceSpans().At(0).Resource()
```

Comment:
Seems a bit presumptuous that the first ResourceSpan of the first set of traces seen is the authoritative source for the host id.

Reply:
Yeah it's pretty hacky, we do the same thing in the signalfxexporter. The logic for determining cloud & host metadata is done in the resourcedetection processor which adds it to the resource. We could export that logic in resourcedetection processor or put it in common but now you would have to set cloud type in multiple places. Maybe there could be an extension but it feels really complicated.